include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp

```diff
@@ -0,0 +1,453 @@
+        if (boost::geometry::closure<ring_2d>::value == boost::geometry::closed)
+        {
+            ring.pop_back(); // Remove the closing point to make it open
+        }
+
+        for (auto it = boost::begin(ring); it != boost::end(ring); ++it)
+        {
+            if (boost::geometry::equals(*it, point1))
+            {
+                if (boost::geometry::equals(point1, point2))
+                {
+                    visitor.template apply<failure_duplicate_points>();
+                    return false;
+                }
+                auto prev = (it == boost::begin(ring)) ? boost::end(ring) - 1 : boost::prior(it);
+                if (boost::geometry::equals(point2, *prev))
+                {
+                    // valid intersection and consistent orientation
+                    return true;
+                }
+                auto next = (it == boost::end(ring)) ? boost::begin(ring) : boost::next(it);
+                if (boost::geometry::equals(point2, *next))
+                {
+                    // valid intersection but inconsistent orientation
+                    if (AllowInconsistentOrientation)
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        visitor.template apply<failure_inconsistent_orientation>();
+                        return false;
+                    }
+                }
+            }
+        }
+        visitor.template apply<failure_invalid_intersection>();
+        return false;
+    }
+
+    // Check if a segment intersects a polygon in 3D space
+    template <typename VisitPolicy, typename Strategy>
+    intersection_type check_polygon_segment_intersection(point_3d const& point1,
+                                                         point_3d const& point2,
+                                                         polygon_3d const& polygon,
+                                                         size_t const& polygon_index,
+                                                         VisitPolicy& visitor,
+                                                         Strategy const& strategy)
+    {
+        vec3 p1;
+        vec3 p2;
+        set_vec_from_point::apply(point1, p1);
+        set_vec_from_point::apply(point2, p2);
+
+        vec3 base;
+        vec3 v1;
+        vec3 v2;
+        vec3 normal;
+        polygon_2d projected_polygon;
+
+        // Check if the polygon has been projected before
+        if (m_polygon_projections[polygon_index].projected_polygon.outer().size() > 0)
+        {
+            auto const& projection = m_polygon_projections[polygon_index];
+            base = projection.base;
+            v1 = projection.v1;
+            v2 = projection.v2;
+            normal = projection.normal;
+            projected_polygon = projection.projected_polygon;
+        }
+        else
+        {
+            auto const& ring = boost::geometry::exterior_ring(polygon);
```

Comment:
So this block does really the projection. Nice.
I think it's good to extract it from this class - and even to another source.
Such that it is useful on its own, and can be unit tested on its own.

Reply:
But exactly that (shared variables) will makes it harder to extract it later as well...

A generic polygon projection from 3d to 2d plane would be great. We could then immediately implement the surface-area algorithm on a 3d surface... 

It would be a generic algorithm, every patch (surface) would be a separate 2D polygon. Or is it different than what you do here (I couldn't look into all the details yet).

Anyway - I'm fine having it as a next PR - I'm only afraid that it would not really safe time.